lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -7190,44 +6873,37 @@ class PullbackEmitter final : public SILInstructionVisitor<PullbackEmitter> {
   ///                            index corresponding to n
   void visitTupleExtractInst(TupleExtractInst *tei) {
     auto *bb = tei->getParent();
+    auto loc = tei->getLoc();
+    auto adjBuf = getAdjointBuffer(bb, tei);
+
+    auto tupleTy = tei->getTupleType();
     auto tupleTanTy = getRemappedTangentType(tei->getOperand()->getType());
-    auto av = getAdjointValue(bb, tei);
-    switch (av.getKind()) {
-    case AdjointValueKind::Zero:
-      addAdjointValue(bb, tei->getOperand(), makeZeroAdjointValue(tupleTanTy),
-                      tei->getLoc());
-      break;
-    case AdjointValueKind::Aggregate:
-    case AdjointValueKind::Concrete: {
-      auto tupleTy = tei->getTupleType();
-      auto tupleTanTupleTy = tupleTanTy.getAs<TupleType>();
-      if (!tupleTanTupleTy) {
-        addAdjointValue(bb, tei->getOperand(), av, tei->getLoc());
-        break;
-      }
-      SmallVector<AdjointValue, 8> elements;
-      unsigned adjIdx = 0;
-      for (unsigned i : range(tupleTy->getNumElements())) {
-        if (!getTangentSpace(
-                tupleTy->getElement(i).getType()->getCanonicalType()))
-          continue;
-        if (tei->getFieldNo() == i)
-          elements.push_back(av);
-        else
-          elements.push_back(makeZeroAdjointValue(
-              getRemappedTangentType(SILType::getPrimitiveObjectType(
-                  tupleTanTupleTy->getElementType(adjIdx++)
-                      ->getCanonicalType()))));
-      }
-      if (elements.size() == 1) {
-        addAdjointValue(bb, tei->getOperand(), elements.front(), tei->getLoc());
-        break;
-      }
-      addAdjointValue(bb, tei->getOperand(),
-          makeAggregateAdjointValue(tupleTanTy, elements), tei->getLoc());
-      break;
+    auto tupleTanTupleTy = tupleTanTy.getAs<TupleType>();
+    if (!tupleTanTupleTy) {
+      addToAdjointBuffer(bb, tei->getOperand(), adjBuf, loc);
+      return;
     }
+
+    // Accumulate adjoint for the `tuple_extract` operand.
+    auto tmp = builder.createAllocStack(loc, tupleTanTy);
+    unsigned adjIdx = 0;
+    for (unsigned i : range(tupleTy->getNumElements())) {
+      if (!getTangentSpace(
+              tupleTy->getElement(i).getType()->getCanonicalType()))
+        continue;
+      auto tmpElement = builder.createTupleElementAddr(loc, tmp, adjIdx);
+      if (tei->getFieldNo() == i) {
+        builder.createCopyAddr(
```

Comment:
This tmp copy is redundant. Instead, just emit `tuple_element_addr` and call `addToAdjointBuffer` with the address to the element.

Reply:
I see, that makes sense.

After differentiation is moved before ownership eliminator, I'm not able to find a test case that tests `PullbackEmitter` visitor for `tuple_extract`. The old `tuple_extract` reproducer now generates `destructure_tuple`:

```swift
func foo(_ x: Float, _ y: Float) -> (Float, Float) {
  return (x + 1, y + 2)
}
@_silgen_name("bar")
func bar(_ x: Float, _ y: Float) -> Float {
  // Try to test `tuple_extract` instruction.
  // Instead, generates `destructure_tuple` since AD comes before ownership eliminator.
  foo(x, y).1
}
print((0, 1), gradient(at: 3, 3, in: bar))
```

```
[AD] Original bb0: To differentiate or not to differentiate?
[ ]   return %7 : $Float                              // id: %8
[∂]   (%6, %7) = destructure_tuple %5 : $(Float, Float) // user: %8
[∂]   %5 = apply %4(%0, %1) : $@convention(thin) (Float, Float) -> (Float, Float) // user: %6
[ ]   // function_ref foo(_:_:)
  %4 = function_ref @$s5hello3fooySf_SftSf_SftF : $@convention(thin) (Float, Float) -> (Float, Float) // user: %5
[ ]   debug_value %1 : $Float, let, name "y", argno 2 // id: %3
[ ]   debug_value %0 : $Float, let, name "x", argno 1 // id: %2
```

What should we do with the `tuple_extract` visitor? Perhaps we can write a SIL test for it.

Reply:
> What should we do with the tuple_extract visitor? Perhaps we can write a SIL test for it.

Actually, I'm leaning towards deleting `tuple_extract` visitors for differential and pullback generation, since it is not testable.

Reply:
Deleted `tuple_extract` visitors in https://github.com/apple/swift/pull/26709/commits/cce599ddc42d6453522b9b1900fb115b8ee8d4db.

Reply:
I applied the same redundant `tmp` allocation optimize for `struct_extract` in https://github.com/apple/swift/pull/26709/commits/a6988c8f54ec6858da48a7f1fe2b8608368c6432.

Reply:
If you do a grep, you'll find that SILGen does generate `tuple_extract` sometimes. While it's hard to come up with a Swift code that results in `tuple_extract`, I think it's better to write a SIL FileCheck test case that triggers this code path than removing the code.

Reply:
Fair, I'll look into add a `tuple_extract` SIL test today.